File: TREES/Pseudo-Palindromic Paths in a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    int cnt = 0;
    
    void preorder(TreeNode* node,int path)
    {
        if(node != NULL)
        {
            path ^= (1 << node->val);
            
            if(node->left == NULL && node->right == NULL)
            {
                if((path &(path -1)) == 0)
                {
                    ++cnt;
                    
                }
            }
            preorder(node->left,path);
            preorder(node->right,path);
            
        }
    }
    
    int pseudoPalindromicPaths (TreeNode* root) {
        preorder(root,cnt);
        return cnt;
        
    }
};

/*
Time complexity: O(N) since one has to visit each node, check if at most one digit has an odd frequency.

Space complexity: up to O(H) to keep the recursion stack, where HH is a tree height.
*/

File: BIT MANIPULATION/pseudo-palindromic-paths-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        res = 0
        def solve(root,val):
            nonlocal res
            if not root:
                return
            val ^= 1 << root.val
            
            if (not(root.left) and not(root.right)):
                if val == 0 or (val&(val-1) == 0):
                    res += 1
                return
            solve(root.left,val)
            solve(root.right,val)
        
        solve(root,0)
        return  res
            
        
